## 8.1 数据类型模块

### 8.1.1 日期和时间处理

#### （1）`time` 模块

In [2]:
import time

# 获取当前时间戳
timestamp = time.time()
print("当前时间戳:", timestamp)

当前时间戳: 1735392062.073334


In [3]:
# 将时间戳转换为可读时间
readable_time = time.ctime(timestamp)
print("可读的当前时间:", readable_time)

可读的当前时间: Sat Dec 28 21:21:02 2024


In [4]:
# 使用 strftime 格式化时间
formatted_time = time.strftime("%Y-%m-%d %H:%M:%S", time.localtime())
print("格式化的当前时间:", formatted_time)

格式化的当前时间: 2024-12-28 21:21:26


In [5]:
# 程序暂停2秒
print("程序暂停2秒...")
time.sleep(2)
print("继续执行")

程序暂停2秒...
继续执行


In [7]:
import time

def progress_bar(total):
    # 开始计时
    start_time = time.perf_counter()
    
    for i in range(total + 1):
        # 计算进度
        percent = (i / total) * 100
        bar_length = 50  # 进度条长度
        block = int(bar_length * i / total)  # 当前已完成的块数
        progress = '#' * block + '-' * (bar_length - block)  # 生成进度条
        print(f'\r[{progress}] {percent:.2f}% 完成', end='')
        time.sleep(0.1)  # 模拟一些工作

    # 结束计时
    end_time = time.perf_counter()
    elapsed_time = end_time - start_time
    print(f'\n总耗时: {elapsed_time:.2f}秒')

# 调用函数，设置总任务数为100
progress_bar(100)

[##################################################] 100.00% 完成
总耗时: 10.45秒


#### （2）`datetime` 模块

In [8]:
from datetime import date, datetime

# 获取当前日期和时间
current_date = date.today()
print("今天的日期:", current_date)
print("今天是星期", current_date.weekday() + 1) 

今天的日期: 2024-12-28
今天是星期 6


In [10]:
# 自定义日期
custom_date = date(2023, 7, 16)
print("自定义日期:", custom_date)

自定义日期: 2023-07-16


In [11]:
# 获取当前时间
current_time = datetime.now().time()
print("当前时间:", current_time)

当前时间: 22:03:42.558044


In [12]:
# 格式化日期时间
now = datetime.now()
formatted_date = now.strftime("%Y-%m-%d %H:%M:%S")
print("格式化的当前日期时间:", formatted_date)

格式化的当前日期时间: 2024-12-28 22:03:53


In [13]:
# 将字符串解析为日期时间
parsed_date = datetime.strptime("2023-10-25", "%Y-%m-%d")
print("解析后的日期:", parsed_date)

解析后的日期: 2023-10-25 00:00:00


In [14]:
from datetime import datetime, timedelta   

# 两个不同的日期
date1 = datetime(2024, 9, 8)
date2 = datetime(2025, 1, 25)

# 计算时间差
difference = date2 - date1
print(f"本学期从 {date1.date()} 开始，{date2.date()} 结束，一共 {difference.days} 天。")

本学期从 2024-09-08 开始，2025-01-25 结束，一共 139 天。


#### （3）`calendar` 模块

In [15]:
import calendar

# 设置星期日为一周的第一天
calendar.setfirstweekday(calendar.SUNDAY)

In [16]:
# 输出某个月的日历
print(calendar.month(2024, 11))

   November 2024
Su Mo Tu We Th Fr Sa
                1  2
 3  4  5  6  7  8  9
10 11 12 13 14 15 16
17 18 19 20 21 22 23
24 25 26 27 28 29 30



In [17]:
# 判断 2024 年是否为闰年
print(calendar.isleap(2024))  # 输出：True

True


In [19]:
# 获取 2024 年 11 月 1 日是星期几
print(calendar.weekday(2024, 11, 1))  # 输出：4（星期五）

# 输出每周的星期标题（3个字符宽度）
print(calendar.weekheader(3))  # 输出：Sun Mon Tue Wed Thu Fri Sat

4
Sun Mon Tue Wed Thu Fri Sat


In [20]:
# 创建 HTMLCalendar 实例
html_cal = calendar.HTMLCalendar(firstweekday=calendar.SUNDAY)

# 生成 HTML 格式的月历
html_month = html_cal.formatmonth(2025, 1)
print(html_month)

<table border="0" cellpadding="0" cellspacing="0" class="month">
<tr><th colspan="7" class="month">January 2025</th></tr>
<tr><th class="sun">Sun</th><th class="mon">Mon</th><th class="tue">Tue</th><th class="wed">Wed</th><th class="thu">Thu</th><th class="fri">Fri</th><th class="sat">Sat</th></tr>
<tr><td class="noday">&nbsp;</td><td class="noday">&nbsp;</td><td class="noday">&nbsp;</td><td class="wed">1</td><td class="thu">2</td><td class="fri">3</td><td class="sat">4</td></tr>
<tr><td class="sun">5</td><td class="mon">6</td><td class="tue">7</td><td class="wed">8</td><td class="thu">9</td><td class="fri">10</td><td class="sat">11</td></tr>
<tr><td class="sun">12</td><td class="mon">13</td><td class="tue">14</td><td class="wed">15</td><td class="thu">16</td><td class="fri">17</td><td class="sat">18</td></tr>
<tr><td class="sun">19</td><td class="mon">20</td><td class="tue">21</td><td class="wed">22</td><td class="thu">23</td><td class="fri">24</td><td class="sat">25</td></tr>
<tr><td

### 8.1.2 双端队列

In [21]:
from collections import deque

# 创建带有元素的 deque
dq = deque([1, 2, 3])

# 添加元素
dq.append(4)         # 右侧添加 4，dq 变为 [1, 2, 3, 4]
dq.appendleft(0)     # 左侧添加 0，dq 变为 [0, 1, 2, 3, 4]

# 移除元素
dq.pop()             # 移除右侧元素 4，dq 变为 [0, 1, 2, 3]
dq.popleft()         # 移除左侧元素 0，dq 变为 [1, 2, 3]

# 扩展 deque
dq.extend([4, 5, 6])        # 右侧添加 [4, 5, 6]，dq 变为 [1, 2, 3, 4, 5, 6]
dq.extendleft([-1, -2, -3]) # 左侧添加 [-1, -2, -3]（顺序反转），dq 变为 [-3, -2, -1, 1, 2, 3, 4, 5, 6]

# 旋转 deque
dq.rotate(2)         # 向右旋转 2 步，dq 变为 [5, 6, -3, -2, -1, 1, 2, 3, 4]
dq.rotate(-1)        # 向左旋转 1 步，dq 变为 [6, -3, -2, -1, 1, 2, 3, 4, 5]

In [22]:
def is_balanced(expression):
    stack = deque()
    for char in expression:
        if char in "({[":
            stack.append(char)
        elif char in ")}]":
            if not stack:
                return False
            top = stack.pop()
            if (top == '(' and char != ')') or (top == '{' and char != '}')\
                  or (top == '[' and char != ']'):
                return False
    return not stack

# 测试
print(is_balanced("(a + b) * [c / d]"))  # 输出 True
print(is_balanced("{a + [b * (c + d)]")) # 输出 False

True
False


### 8.1.3 枚举类型

In [23]:
from enum import Enum

class Color(Enum):
    RED = 1
    GREEN = 2
    BLUE = 3

In [24]:
# 通过名称访问
print(Color.RED)         # 输出：Color.RED
print(Color.RED.name)    # 输出：RED
print(Color.RED.value)   # 输出：1

# 通过值访问
color = Color(2)
print(color)             # 输出：Color.GREEN

Color.RED
RED
1
Color.GREEN


In [25]:
for color in Color:
    print(color)    # 输出：Color.RED, Color.GREEN, Color.BLUE

Color.RED
Color.GREEN
Color.BLUE


## 8.2 数字和数学运算

### 8.2.1 math 模块

#### （1）常用常数

In [26]:
import math     #导入模块

print(math.pi)  #输出math.pi
print(math.e)   #输出math.e
print(math.tau) #输出math.tau
print(math.inf) #输出math.inf
print(math.nan) #输出math.nan

3.141592653589793
2.718281828459045
6.283185307179586
inf
nan


#### （2）基本函数

- **数论与表示函数**

In [27]:
import math
a, b = 36, 48

# 1. 输出 a 和 b 的绝对值
abs_a = math.fabs(a)
abs_b = math.fabs(b)
print(f"Absolute values: |{a}| = {abs_a}, |{b}| = {abs_b}")

Absolute values: |36| = 36.0, |48| = 48.0


In [28]:
# 2. 计算并输出最大公约数
gcd_ab = math.gcd(a, b)
print(f"Greatest Common Divisor (GCD) of {a} and {b} is {gcd_ab}")

Greatest Common Divisor (GCD) of 36 and 48 is 12


In [29]:
# 3. 计算并输出最小公倍数
if a != 0 and b != 0:
    lcm_ab = abs(a * b) // gcd_ab
    print(f"Least Common Multiple (LCM) of {a} and {b} is {lcm_ab}")
else:
    print("LCM is not defined for zero values.")

Least Common Multiple (LCM) of 36 and 48 is 144


In [30]:
# 4. 计算 gcd_ab 的阶乘（前提是 gcd_ab 为非负整数）
if gcd_ab >= 0:
    print(f"{gcd_ab}! = {math.factorial(gcd_ab)}")
else:
    print("Factorial is only defined for non-negative integers.")

12! = 479001600


In [31]:
# 5. 计算 a 的整数平方根（非负整数）
if a >= 0:
    isqrt_a = math.isqrt(a)
    print(f"Integer square root of {a} is {isqrt_a}")
else:
    print("Integer square root is only defined for non-negative integers.")

Integer square root of 36 is 6


In [32]:
# 6. 计算 a 的符号与 b 相同的副本
copysign_ab = math.copysign(a, b)
print(f"Value of {a} with the sign of {b} is {copysign_ab}")

Value of 36 with the sign of 48 is 36.0


In [33]:
# 7. 将 a 和 b 拆分成整数和小数部分
modf_a = math.modf(a)
modf_b = math.modf(b)
print(f"{a} split into fractional and integer parts: {modf_a}")
print(f"{b} split into fractional and integer parts: {modf_b}")

36 split into fractional and integer parts: (0.0, 36.0)
48 split into fractional and integer parts: (0.0, 48.0)


- **幂函数与对数函数**

In [34]:
import math

# 初始投资金额1000, 年利率5%, 投资年限10年
P, r, t = 1000, 0.05, 10 

# 计算复利总金额
A = P * math.pow(1 + r, t)
print(f"经过 {t} 年后的总金额为：{A:.2f} 元")

经过 10 年后的总金额为：1628.89 元


In [35]:
# 计算翻倍所需时间
t_double = math.log(2) / math.log(1 + r)
print(f"在年利率 {r * 100}% 下，投资翻倍所需时间为：{t_double:.2f} 年")

在年利率 5.0% 下，投资翻倍所需时间为：14.21 年


- **三角函数**

In [36]:
import math
# 已知的直角边和斜边
a = 3   # 直角边 a 的长度
c = 5   # 斜边 c 的长度

# 计算另一条直角边 b 的长度
b = math.sqrt(math.pow(c, 2) - math.pow(a, 2))
print(f"另一条直角边 b 的长度为：{b:.2f}")

另一条直角边 b 的长度为：4.00


In [37]:
# 计算角 A 的弧度值
angle_A_radians = math.asin(a / c)
# 将角 A 转换为度数
angle_A_degrees = math.degrees(angle_A_radians)
print(f"角 A 的大小为：{angle_A_degrees:.2f} 度")

角 A 的大小为：36.87 度


In [38]:
# 计算角 B 的大小
angle_B_degrees = 90 - angle_A_degrees
print(f"角 B 的大小为：{angle_B_degrees:.2f} 度")

角 B 的大小为：53.13 度


In [40]:
# 验证正弦和余弦的计算
sin_A = math.sin(angle_A_radians)  # 正弦
cos_A = math.cos(angle_A_radians)  # 余弦
tan_A = math.tan(angle_A_radians)  # 正切

print(f"角 A 的正弦值：{sin_A:.2f}")
print(f"角 A 的余弦值：{cos_A:.2f}")
print(f"角 A 的正切值：{tan_A:.2f}")

角 A 的正弦值：0.60
角 A 的余弦值：0.80
角 A 的正切值：0.75


- **组合与排列函数**

In [41]:
import math

# 总数和选择数
total_numbers = 49  # 可选号码总数
choose_numbers = 6  # 每次选择的号码数量

# 计算组合数
total_combinations = math.comb(total_numbers, choose_numbers)
print(f"从49个号码中选择6个号码的组合数为：{total_combinations}")

# 计算排列数
total_permutations = math.perm(total_numbers, choose_numbers)
print(f"从49个号码中选择6个号码的排列数为：{total_permutations}")

从49个号码中选择6个号码的组合数为：13983816
从49个号码中选择6个号码的排列数为：10068347520


### 8.2.2 cmath 模块

In [42]:
import cmath

# 定义一个复数
z = 3 + 4j  # 复数 z = 3 + 4i
# 计算复数的平方根
sqrt_z = cmath.sqrt(z)
print(f"复数 z: {z}")
print(f"z 的平方根: {sqrt_z}")

复数 z: (3+4j)
z 的平方根: (2+1j)


In [43]:
# 计算复数的自然对数
log_z = cmath.log(z)
print(f"z 的自然对数: {log_z}")

z 的自然对数: (1.6094379124341003+0.9272952180016122j)


In [44]:
# 计算复数的正弦
sin_z = cmath.sin(z)
# 计算复数的余弦
cos_z = cmath.cos(z)
print(f"z 的正弦: {sin_z}")
print(f"z 的余弦: {cos_z}")

z 的正弦: (3.853738037919377-27.016813258003936j)
z 的余弦: (-27.034945603074224-3.8511533348117775j)


In [45]:
# 展示正虚无穷、非数字
print(f"正虚无穷: {cmath.infj}")
print(f"非数字: {cmath.nanj}")

正虚无穷: infj
非数字: nanj


### 8.2.3 decimal 模块

In [46]:
from decimal import Decimal, getcontext

# 使用float进行计算
float_result = 0.1 + 0.2
# 使用decimal进行高精度计算
getcontext().prec = 28  # 设置decimal精度为28位
decimal_result = Decimal('0.1') + Decimal('0.2')

# 打印结果
print(f"使用 float 计算的结果: {float_result}")
print(f"使用 decimal 计算的结果: {decimal_result}")
# 打印结果的差异
print(f"float 与 decimal 的差异: {float_result - float(decimal_result)}")

使用 float 计算的结果: 0.30000000000000004
使用 decimal 计算的结果: 0.3
float 与 decimal 的差异: 5.551115123125783e-17


### 8.2.4 fractions 模块

In [47]:
from fractions import Fraction

# 创建分数
f1 = Fraction(1, 3)   # 1/3
f2 = Fraction(3, 4)   # 3/4
f3 = Fraction('2/5')  # 2/5
# 打印分数
print(f"f1: {f1}, f2: {f2}, f3: {f3}")

f1: 1/3, f2: 3/4, f3: 2/5


In [48]:
# 基本运算
sum_result = f1 + f2    # 1/3 + 3/4
difference_result = f2 - f1  # 3/4 - 1/3
product_result = f1 * f3    # 1/3 * 2/5
quotient_result = f2 / f1   # 3/4 / 1/3

print(f"Sum: {sum_result}")     # 13/12
print(f"Difference: {difference_result}")  # 5/12
print(f"Product: {product_result}")     # 2/15
print(f"Quotient: {quotient_result}")   # 9/4

Sum: 13/12
Difference: 5/12
Product: 2/15
Quotient: 9/4


In [49]:
# 获取分子和分母
print(f"Numerator of f2: {f2.numerator}")  # 3
print(f"Denominator of f2: {f2.denominator}")  # 4

Numerator of f2: 3
Denominator of f2: 4


In [50]:
# 约简分数
f4 = Fraction(10, 100)  # 10/100
print(f"Reduced fraction of 10/100: {f4}")  # 1/10

Reduced fraction of 10/100: 1/10


In [51]:
# 查找最接近的分数, 寻找1/3最接近的分数，分母不超过10
closest_fraction = f1.limit_denominator(10) 
print(f"Closest fraction to 1/3 with denominator <= 10: {closest_fraction}")

Closest fraction to 1/3 with denominator <= 10: 1/3


### 8.2.5 random 模块

In [52]:
import random

# 生成一个随机浮点数
random_float = random.random()
print(f"随机浮点数: {random_float}")

随机浮点数: 0.2610383239320385


In [53]:
# 生成一个随机整数
random_integer = random.randint(1, 10)  # 从1到10（包含1和10）
print(f"随机整数: {random_integer}")

随机整数: 8


In [54]:
# 生成一个随机浮点数在指定范围内
random_uniform = random.uniform(1.0, 10.0)  # 从1.0到10.0（不包括10.0）
print(f"随机浮点数（范围内）: {random_uniform}")

随机浮点数（范围内）: 2.7837184073130503


In [55]:
# 从列表中随机选择一个元素
options = ['红', '蓝', '绿', '黄']
random_choice = random.choice(options)
print(f"随机选择的颜色: {random_choice}")

随机选择的颜色: 蓝


In [56]:
# 从列表中随机选择多个元素
random_sample = random.sample(options, 2)  # 随机选择两个不同的颜色
print(f"随机选择的颜色样本: {random_sample}")

随机选择的颜色样本: ['蓝', '红']


In [57]:
# 打乱列表中的元素
random.shuffle(options)
print(f"打乱后的颜色列表: {options}")

打乱后的颜色列表: ['蓝', '红', '黄', '绿']


In [58]:
# 设置随机种子
random.seed(42)  # 每次使用相同的种子生成相同的随机数
print("使用种子42生成的随机数：",random.random())

使用种子42生成的随机数： 0.6394267984578837


In [59]:
# 使用 uniform() 生成一个范围在 [1.0, 10.0] 之间的随机浮点数
uniform_random = random.uniform(1.0, 10.0)
print(f"均匀分布的随机浮点数 (1.0 to 10.0): {uniform_random}")

均匀分布的随机浮点数 (1.0 to 10.0): 1.2250967970040025


In [60]:
# 使用 gauss() 生成一个以 0 为均值，1 为标准差的高斯分布随机数
gauss_random = random.gauss(0, 1)
print(f"高斯分布随机数 (mean=0, std=1): {gauss_random}")

高斯分布随机数 (mean=0, std=1): -0.11131586156766247


In [61]:
# 使用 expovariate() 生成一个负指数分布随机数，参数为 λ=1.0
expovariate_random = random.expovariate(1.0)
print(f"负指数分布随机数 (lambda=1.0): {expovariate_random}")

负指数分布随机数 (lambda=1.0): 1.333592672808083


## 8.3 绘图模块

### 8.3.3 turtle 模块的应用

#### （1）使用 `turtle` 模块绘图

In [76]:
import turtle

# 直接调用turtle函数控制默认海龟
turtle.setup(300,300) # 设置屏幕的宽度和高度
turtle.forward(100)   # 向前移动100像素
turtle.right(90)      # 右转90度
turtle.forward(100)   # 向前移动100像素
turtle.right(90)      # 右转90度
turtle.forward(100)   # 向前移动100像素
turtle.right(90)      # 右转90度
turtle.forward(100)   # 向前移动100像素

turtle.done()         # 结束绘图，保持窗口打开

#### （2）构建 `Turtle` 对象绘图

In [84]:
import turtle

turtle.setup(350,300)      # 设置默认screen的宽度和高度
# 创建第一个Turtle对象，用于绘制正方形
square_turtle = turtle.Turtle()
square_turtle.color("blue")      # 设置海龟的颜色为蓝色
square_turtle.pensize(2)         # 设置线宽

# 绘制正方形
for _ in range(4):
    square_turtle.forward(100)   # 向前移动100像素
    square_turtle.right(90)      # 右转90度

# 创建第二个Turtle对象，用于绘制三角形
triangle_turtle = turtle.Turtle()
triangle_turtle.color("red")     # 设置海龟的颜色为红色
triangle_turtle.pensize(2)       # 设置线宽
triangle_turtle.penup()          # 提起画笔，不绘制路径
triangle_turtle.goto(-150, 0)    # 移动到新位置
triangle_turtle.pendown()        # 放下画笔

# 绘制三角形
for _ in range(3):
    triangle_turtle.forward(100) # 向前移动100像素
    triangle_turtle.left(120)    # 左转120度

turtle.done()                    # 保持窗口打开

2024-12-28 23:11:32.551 python[66363:6454079] _TIPropertyValueIsValid called with 12 on nil context!
2024-12-28 23:11:32.551 python[66363:6454079] imkxpc_setApplicationProperty:value:reply: called with incorrect property value 12, bailing.


#### （3）构建 `RawTurtle` 对象绘图

In [86]:
import turtle

# Step 1: 创建一个 TurtleScreen 实例作为绘图窗口
screen = turtle.getscreen()         # 创建一个 TurtleScreen 实例
screen.title("Using RawTurtle Example")  # 设置窗口标题
screen.bgcolor("lightyellow")       # 设置背景颜色
screen.setup(350,300,0,0)               # 设置窗口宽度和高度

# Step 2: 在 TurtleScreen 实例上创建一个 RawTurtle 对象
raw_turtle = turtle.RawTurtle(screen)
raw_turtle.color("purple")          # 设置海龟颜色
raw_turtle.pensize(3)               # 设置画笔宽度

# Step 3: 使用 RawTurtle 对象绘图
for _ in range(4):
    raw_turtle.forward(100)         # 向前移动100像素
    raw_turtle.right(90)            # 右转90度，形成一个正方形

# 保持窗口打开，直到用户关闭窗口
turtle.done()

#### （4）`turtle` 绘图的综合示例

## 8.4 系统与环境管理模块

### 8.4.1 os 模块

In [69]:
import os

# 获取当前工作目录
current_directory = os.getcwd()
print("当前工作目录:", current_directory)

# 列出当前目录中的文件和目录
print("当前目录内容:", os.listdir())

# 创建目录
os.mkdir("test_directory")
print("创建了新的目录 'test_directory'")

# 删除目录
os.rmdir("test_directory")
print("删除了目录 'test_directory'")

# 获取文件的绝对路径
file_path = os.path.abspath("example.txt")
print("文件的绝对路径:", file_path)

# 检查路径是否存在
if os.path.exists(file_path):
    print("路径存在:", file_path)
else:
    print("路径不存在:", file_path)

# 检查路径是否为文件或目录
print("是否为文件:", os.path.isfile(file_path))
print("是否为目录:", os.path.isdir(file_path))

# 获取环境变量
path_variable = os.getenv('PATH')
print("环境变量 PATH:", path_variable)

# 运行系统命令
os.system('echo Hello from os.system')  # 执行系统命令

当前工作目录: /Users/yif/Desktop/Python第2版/JupyterCode25/chapter_8_code
当前目录内容: ['creating_RawTurtle_obj.py', '.DS_Store', 'comprehensive_turtle_exam.py', 'using_turtle.py', '实验7 特殊方法和属性.ipynb', 'draw_date_turtle.py', 'landscape.gif', 'creating_Turtle_obj.py', '.ipynb_checkpoints', 'chapter_8_code.ipynb']
创建了新的目录 'test_directory'
删除了目录 'test_directory'
文件的绝对路径: /Users/yif/Desktop/Python第2版/JupyterCode25/chapter_8_code/example.txt
路径不存在: /Users/yif/Desktop/Python第2版/JupyterCode25/chapter_8_code/example.txt
是否为文件: False
是否为目录: False
环境变量 PATH: /Users/yif/anaconda3/bin:/opt/homebrew/bin:/opt/homebrew/sbin:/usr/local/bin:/System/Cryptexes/App/usr/bin:/usr/bin:/bin:/usr/sbin:/sbin:/var/run/com.apple.security.cryptexd/codex.system/bootstrap/usr/local/bin:/var/run/com.apple.security.cryptexd/codex.system/bootstrap/usr/bin:/var/run/com.apple.security.cryptexd/codex.system/bootstrap/usr/appleinternal/bin
Hello from os.system


0

### 8.4.2 sys 模块

In [70]:
import sys

# 获取Python解释器的版本信息
print("Python Version:", sys.version)

# 获取命令行参数列表
print("Command Line Arguments:", sys.argv)

# 获取模块搜索路径
print("Module Search Path:", sys.path)

# 获取Python支持的最大整数
print("The Maximal Integer", sys.maxsize)

# 退出程序，状态码为0表示正常退出
sys.exit(0)

Python Version: 3.10.9 (main, Mar  1 2023, 12:33:47) [Clang 14.0.6 ]
Command Line Arguments: ['/Users/yif/anaconda3/lib/python3.10/site-packages/ipykernel_launcher.py', '-f', '/Users/yif/Library/Jupyter/runtime/kernel-20a58921-c0dc-4a50-8146-5631c70a903b.json']
Module Search Path: ['/Users/yif/Desktop/Python第2版/JupyterCode25/chapter_8_code', '/Users/yif/anaconda3/lib/python310.zip', '/Users/yif/anaconda3/lib/python3.10', '/Users/yif/anaconda3/lib/python3.10/lib-dynload', '', '/Users/yif/anaconda3/lib/python3.10/site-packages', '/Users/yif/anaconda3/lib/python3.10/site-packages/PyQt5_sip-12.11.0-py3.10-macosx-10.9-x86_64.egg', '/Users/yif/anaconda3/lib/python3.10/site-packages/aeosa', '/Users/yif/anaconda3/lib/python3.10/site-packages/mpmath-1.2.1-py3.10.egg']
The Maximal Integer 9223372036854775807


SystemExit: 0

/Users/yif/anaconda3/lib/python3.10/site-packages/IPython/core/interactiveshell.py:3468: UserWarning: To exit: use 'exit', 'quit', or Ctrl-D.
  warn("To exit: use 'exit', 'quit', or Ctrl-D.", stacklevel=1)


## 8.5 上机实践

1. 请使用`datetime`模块计算用户年龄（由用户输入生日），判断用户的年龄类别。

	&emsp;&emsp;拓展资料：根据 2017 年联合国世界卫生组织发布的年龄分段标准进行分类：\\(44\\) 岁及以下为青年人，\\(45\\) 岁到 \\(59\\) 岁为中年人，\\(60\\) 岁到 \\(74\\) 岁为年轻老年人，\\(75\\) 岁到 \\(89\\) 岁为老年人，\\(90\\) 岁及以上为长寿老年人。 

In [ ]:
from datetime import datetime

# 定义年龄分类的边界
youth_limit = 44
middle_age_start = 45
middle_age_end = 59
young_senior_start = 60
senior_start = 75
longevity_start = 90

# 主程序
birth_date_str = input("请输入您的出生日期（YYYY-MM-DD）：")
birth_date = datetime.strptime(birth_date_str, "%Y-%m-%d")  # 将输入的字符串转换为datetime对象

# 计算年龄
today = datetime.now()
age = today.year - birth_date.year

# 如果当前的月份和日期小于出生的月份和日期, 则 age - 1
if (today.month, today.day) < (birth_date.month, birth_date.day):
    age -= 1

# 判断年龄类别
if age <= youth_limit:
    category = "青年人"
elif middle_age_start <= age <= middle_age_end:
    category = "中年人"
elif young_senior_start <= age <= senior_start:
    category = "年轻老年人"
elif senior_start <= age <= longevity_start:
    category = "老年人"
else:
    category = "长寿老年人"

print(f"您的年龄是：{age} 岁，分类为：{category}")

2. 使用枚举`Enum`类型定义扑克牌的花色（红心、方块、黑桃、梅花）和面值（2至A），并实现下列功能：
	- 定义`create_deck()`函数生成一副完整的扑克牌，牌面用元组的形式存储。
	- 定义`shuffle_deck(deck, seed)`函数洗牌，用户输入一个“幸运数”作为随机种子`seed`。
	- 定义`draw_cards(deck, num=5)`函数从洗好的牌组中随机抽取5张牌。
	- 如果抽出的牌中由一对，则输出“You are luck!”，否则，输出"Just no bad."

In [5]:
import random
from enum import Enum

class Suit(Enum):
    HEARTS, DIAMONDS, SPADES, CLUBS = "红心", "方块", "黑桃", "梅花"

class Rank(Enum):
    TWO,THREE,FOUR,FIVE = "2","3","4","5"
    SIX,SEVEN,EIGHT,NINE,TEN = "6","7","8","9","10"
    JACK = "J"
    QUEEN = "Q"
    KING = "K"
    ACE = "A"

# 创建扑克牌
def create_deck():
    deck = []
    for suit in Suit:
        for rank in Rank:
            card = (rank.value, suit.value)  # 使用元组存储牌的面值和花色
            deck.append(card)
    return deck

# 洗牌
def shuffle_deck(deck, seed):
    random.seed(seed)  # 设置随机数种子
    random.shuffle(deck)

# 抽取5张牌
def draw_cards(deck, num=5):
    return random.sample(deck, num)

# 检查是否有一对
def has_pair(hand):
    ranks = [card[0] for card in hand]   # 提取牌的面值
    return len(set(ranks)) < len(ranks)  # 判断是否有重复的面值

# 主程序
if __name__ == "__main__":
    lucky_number = int(input("请输入一个幸运数: "))
    deck = create_deck()         # 创建一副扑克牌
    shuffle_deck(deck, lucky_number)  # 使用用户的幸运数洗牌
    hand = draw_cards(deck, 5)  # 随机抽取5张牌
    
    # 输出抽到的牌
    print("随机抽取的5张牌:", end = '\t')
    for card in hand:
        print(f"{card[1]} {card[0]}", end = '\t')
    print()

# 检查是否有一对
    if has_pair(hand):
        print("You are luck!")
    else:
        print("Just no bad.")

请输入一个幸运数: 5
随机抽取的5张牌:	红心 A	方块 9	方块 7	黑桃 2	红心 7	
You are luck!


3. 蒙特卡洛方法是一种利用随机数（通常是伪随机数）来解决计算问题的数值方法。通过随机撒点并观察点的分布情况，可以计算出面积的比值，进而估算圆周率的近似值。这种方法体现了Python在数学问题中运用计算思维的能力。

	&emsp;&emsp;请使用蒙特卡洛方法计算圆周率的近似值，上机任务要求：
	
	- 在一个边长为2的正方形内随机撒点（即，生成点的坐标在区间\\([-1, 1]\\)范围内）。
	- 计算这些点中落在单位圆内的比例（单位圆的圆心在坐标原点，半径为 \\(1\\)），并利用该比例来估算圆周率的近似值。
	- 记录并输出程序的执行时间，以评估计算效率。

In [6]:
import random, math, time

# 记录程序开始运行时间
start_time = time.perf_counter()

# 设置循环次数（即撒点数量），为100万个点
total_points = 1_000_000

# 初始化落入圆中的点的计数器
circle_points = 0

# 循环撒点
for _ in range(total_points):
    # 生成随机点的x和y坐标，范围为[0, 1)
    x = random.random()
    y = random.random()  
    
    # 计算点到原点的距离
    distance = math.sqrt(x**2 + y**2)
    
    # 如果点在单位圆内（距离小于等于1），则计数
    if distance <= 1:
        circle_points += 1

# 利用撒点比例计算圆周率的近似值
estimated_pi = 4 * (circle_points / total_points)
print(f"Estimated π: {estimated_pi}")

# 记录程序结束时间，并输出运行时长
end_time = time.perf_counter()
print(f"Execution time: {end_time - start_time:.2f} seconds")

Estimated π: 3.142564
Execution time: 0.38 seconds


4. 请编写程序实现当前日期的七段晶体管绘制。

In [ ]:
import turtle
import time

# 定义函数 drawGap，用于在绘制线条之间留出5个像素的间隔
def drawGap():
    turtle.penup()
    turtle.fd(5)

# 定义函数 drawLine，根据参数决定是否绘制一段长度为40的线条
def drawLine(draw):
    drawGap()
    turtle.pendown() if draw else turtle.penup()
    turtle.fd(40)
    drawGap()
    turtle.right(90)

# 定义函数 drawDigit，用于绘制单个数字的7段数码管样式
def drawDigit(digit):
    drawLine(True) if digit in [2, 3, 4, 5, 6, 8, 9] else drawLine(False)
    drawLine(True) if digit in [0, 1, 3, 4, 5, 6, 7, 8, 9] else drawLine(False)
    drawLine(True) if digit in [0, 2, 3, 5, 6, 8, 9] else drawLine(False)
    drawLine(True) if digit in [0, 2, 6, 8] else drawLine(False)
    turtle.left(90)
    drawLine(True) if digit in [0, 4, 5, 6, 8, 9] else drawLine(False)
    drawLine(True) if digit in [0, 2, 3, 5, 6, 7, 8, 9] else drawLine(False)
    drawLine(True) if digit in [0, 1, 2, 3, 4, 7, 8, 9] else drawLine(False)
    turtle.left(180)

    turtle.penup()
    turtle.fd(20)  # 前进数字之间的分隔20像素

# 定义函数 drawDate，用于绘制日期，并根据字符内容设置画笔颜色
def drawDate(date):
    turtle.pencolor("red")  # 初始化画笔颜色

    for char in date:
        if char == '-':
            turtle.right(90)
            turtle.fd(20)   # 绘笔下移，文字在中间显示
            turtle.left(90)
            turtle.write("年", font=("Simsum", 38, "normal"))
            turtle.pencolor("green")
            turtle.left(90)
            turtle.fd(20)   # 绘笔回归原位
            turtle.right(90)
            turtle.fd(40)

        elif char == '=':
            turtle.right(90)
            turtle.fd(20)   #  绘笔下移，文字在中间显示
            turtle.left(90)
            turtle.write("月", font=("Simsum", 38, "normal"))
            turtle.pencolor("blue")
            turtle.left(90)
            turtle.fd(20)   # 绘笔回归原位
            turtle.right(90)
            turtle.fd(40)

        elif char == '+':
            turtle.right(90)
            turtle.fd(20)   # 文字在中间显示
            turtle.left(90)
            turtle.write("日", font=("Simsum", 38, "normal"))

        else:
            drawDigit(eval(char))  # 绘制数字字符

# 主程序
if __name__ == "__main__":
    turtle.setup(800, 350)  # 定义绘图窗口大小
    turtle.penup()
    turtle.fd(-350)
    turtle.pensize(5)    # 设置画笔粗细
    drawDate(time.strftime("%Y-%m=%d+", time.gmtime()))  # 绘制当前日期
    turtle.hideturtle()  # 隐藏海龟
    turtle.done()	# 保持窗口打开

## 上机作业

1. 获取当前日期时间，并输出当月的日历。

2. 将浮点数 $3.14$ 和 `decimal` 对象 `Decimal(3.14)`转换为 `Fraction` 实例，并求该实例的向上取 整值和向下取整值、舍入值以及近似估计值。

3. 使用 turtle 模块绘制一个红色五角星。

4. **（自选题）** 随机生成1000个符合正太分布的浮点数（均值为5，标准差为1），并从这些生成的数字中随机再随机选择100个，对这100个数字进行四舍五入，再统计操作后各个数字出现的次数。使用*数目表示不同数字的出现次数，并按数字从小到大打印出来，统计程序运行时间，观察图形分布。